### PR TITLE
Feature/issue 456 sub missing latlong

### PR DESCRIPTION
**Pull request by ian-lin-PG**

### CODE REVIEW

File: bc211/management/commands/import_bc211_data.py [hunk 1]
```diff
@@ -17,23 +18,32 @@ def add_arguments(self, parser):
                             type=argparse.FileType('r'),
                             metavar='file',
                             help='Path to XML file containing BC-211 data')
+        parser.add_argument('--csv',
+                            metavar='csv',
+                            help='Path to CSV file containing city to latlong dictionary')
```
Comment (rasmus-storjohann-PG) [inline bc211/management/commands/import_bc211_data.py]: `--city_latlongs`

File: bc211/management/commands/import_bc211_data.py [hunk 2]
```diff
@@ -17,23 +18,32 @@ def add_arguments(self, parser):
...
 
     def handle(self, *args, **options):
         counts = ImportCounters()
         file = options['file']
+        if options['csv']:
+            csv_path = options['csv']
+        else:
+            csv_path = None
```
Comment (rasmus-storjohann-PG) [inline bc211/management/commands/import_bc211_data.py]:
```py
        if options['csv']:
            city_latlongs = parse_csv(options['csv'])
        else:
            city_latlongs = {}
```

File: bc211/importer.py [hunk 3]
```diff
@@ -12,20 +14,27 @@
...
+    with open(csv_path, mode='r') as file:
+        csv_reader = csv.reader(file)
+        city_to_latlong = {rows[0]: Point(float(rows[1]), float(rows[2])) for rows in csv_reader}
+        return city_to_latlong
+
+
 def save_records_to_database(organizations, counters):
     for organization in handle_parser_errors(organizations):
-        save_organization(organization, counters)
+        save_organization(organization, None, counters)
```
Comment (rasmus-storjohann-PG) [inline bc211/importer.py]: pass in `{}`, then you can remove downstream tests against `None`

File: bc211/importer.py [hunk 4]
```diff
@@ -70,6 +80,35 @@ def save_locations(locations, counters):
...
+
+def validate_city_latlong_map(city_latlong_map):
+    if not city_latlong_map:
+        return {
+            'New Westminster':  Point(-122.910956, 49.205718),
+            'Coquitlam':        Point(-122.793206, 49.283763),
+            'Vancouver':        Point(-123.120738, 49.282729),
+            'Delta':            Point(-123.026476, 49.095216),
+        }
+    return city_latlong_map
```
Comment (rasmus-storjohann-PG) [inline bc211/importer.py]: Remove this function

File: bc211/importer.py [hunk 5]
```diff
@@ -70,6 +80,35 @@ def save_locations(locations, counters):
             create_phone_numbers_for_location(active_record, location.phone_numbers, counters)
 
 
+def validate_latlong_on_location(location_dto, city_latlong_map):
+    validated_city_latlong_map = validate_city_latlong_map(city_latlong_map)
+    location_dto_no_latlong = location_dto.spatial_location is None
+    location_has_physical_address = location_dto.physical_address is not None
```
Comment (rasmus-storjohann-PG) [inline bc211/importer.py]:
```py
if location_dto.spatial_location:
    return location_dto
if location_dto.physical_address is None:
    print_warning
    return location_dto
set location ....
```

File: bc211/tests/test_importer.py [hunk 6]
```diff
@@ -41,6 +59,34 @@ def test_can_import_latitude(self):
...
         self.assertAlmostEqual(self.location.point.x, -122.607918)
 
+    def test_can_replace_missing_lat_long_with_city_lat_long(self):
+        organization = OrganizationBuilder().create()
+        location_id = a_string()
+        address_dto = dtos.Address(location_id=location_id,
+                                   address_type_id=self.city_address_with_latlong['address_type_id'],
+                                   city=self.city_address_with_latlong['city'], country=self.city_address_with_latlong['country'],
+                                   address_lines=a_string(),
+                                   state_province=a_string(), postal_code=a_string())
```
Comment (rasmus-storjohann-PG) [inline bc211/tests/test_importer.py]: Use builder here

File: bc211/tests/test_importer.py [hunk 7]
```diff
@@ -41,6 +59,34 @@ def test_can_import_latitude(self):
...
+        location_dto = LocationBuilder(organization).with_id(location_id).with_point(
```
Comment (rasmus-storjohann-PG) [inline bc211/tests/test_importer.py]: location_with_address

File: bc211/importer.py [hunk 8]
```diff
@@ -81,34 +81,25 @@ def save_locations(locations, city_latlong_map, counters):
...
+    if location_dto.physical_address.city in city_latlong_map:
+        replacement_point = city_latlong_map[location_dto.physical_address.city]
+        replacement_spatial_location = dtos.SpatialLocation(
+            latitude=replacement_point.y,
+            longitude=replacement_point.x
+        )
+        location_dto.spatial_location = replacement_spatial_location
+        return location_dto
+    LOGGER.warning('Location with id "%s" does not have city to fall back on for LatLong info',
+                   location_dto.id)
```
[the remark this answers is not in the document]
Comment (ian-lin-PG) [reply bc211/importer.py]: the default case of city not in the city_latlong_map can be return earlier than the latlong replacement case.

[the remark this answers is not in the document]
Comment (rasmus-storjohann-PG) [reply bc211/importer.py]: That is correct

**Pull request merged**